### DISCUSSION TITLE
How to delete a branch through the API?

### DISCUSSION THREAD

**Discussion by pauleustice** [community/community#24603] [Apps, API and Webhooks]
<p>Hi everyone! I’m searching for a way to delete a branch through the API (REST or Octokit). I need to interact with the branch in a workflow after merging the PR (on the pull request close event), and then want to delete it once I’m done. However the documentation makes no mention of this seemingly obvious requirement, so am I missing something?</p>

**Comment by pauleustice** [comment, marked as answer]
<p>I found a solution (which isn’t mentioned in the docs anywhere):</p>
<pre><code class="lang-auto">curl -s -X DELETE -u username:${{secrets.GITHUB_TOKEN}} https://api.github.com/repos/${{ github.repository }}/git/refs/heads/${{ github.head_ref }}
</code></pre>

**Reply by kmason-extron** [reply]
https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#delete-a-reference

**Comment by surajgaud** [comment]
<p>branch can be deleted using references, refer to the references API docs for more info.<br>
for example in python you can use the following,</p>
<pre><code class="lang-auto">requests.delete(f"{API_URL}/repos/{OWNER}/{REPO}/git/refs/heads/{BRANCH_NAME}")
</code></pre>
<p>a successful operation returns 204 code</p>

**Comment by AlexSneedMiller** [comment]
<p>Is there a way to delete a branch via the API that also allows you to later restore the branch if you decide you want it back? When you delete a branch in the UI you can always go back and ‘restore’ it.</p>
<p>When you delete the branch via the API, you can’t “search” for the branch anymore and it doesn’t appear that you can restore it via the UI.</p>

**Reply by AlexSneedMiller** [reply]
Ended up writing a script to delete references. Branches are just references, which are just names of specific commits. As long as the commit is there you can always go back.

**Reply by kmason-extron** [reply]
branches are nothing more than a symbolic reference to a git SHA. That said there are a few ways I have gone about this. Given that creating a branch again only requires the last SHA of the branch at deletion (so you can checkout to the SHA and then make a branch with same or different name) you can:

1. Tag that commit before delete so you have a git reference to go back to
    - Pros: native git reference that is there until you delete tag (though SHA always there unless rewrite history)
    - Cons: Depending on workflow you now just have a mess of tags mixed in with other uses of tags
1. Before delete save the SHA elsewhere
    - You can save in a wiki
    - A file in repo (think changelog etc)
    - Pros: no cluttering of tags, could be more scalable to multiple repos with a wiki etc
    - Cons: Isn't git native per-se. While saving in a file in the repo is git you aren't using git command that makes it easy to undo 
1. Use commit messages for things that make sense like PR merges etc 
    - Pros: native git sorta as you then can search for a message in a given format to find its commit. 
    - Cons: searching commits can be a pain in cli or api
    - Notes: the PR commit is the commit on the merged to branch so you would have to look at the parent(s)

**Reply by kmason-extron** [reply]
> Ended up writing a script to delete references. Branches are just references, which are just names of specific commits. As long as the commit is there you can always go back.

@AlexSneedMiller Sorry my post was about 10 min late.
